File: contrib/pyre/tests/mpi/group_setops.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# michael a.g. aïvázis
# orthologue
# (c) 1998-2019 all rights reserved
#


"""
Exercise the group manipulation interface.
"""


def test():
    # access the package
    import mpi
    # initialize
    mpi.init()
    # grab the world communicator
    world = mpi.world
    # access the world process group
    whole = world.group()
    # build a tuple of the even ranks
    ranks = tuple(rank for rank in range(world.size) if (rank % 2 == 0))

    # build two groups
    odds = whole.exclude(ranks)
    evens = whole.include(ranks)

    # compute the union of the two
    union = odds.union(evens)
    # verify that the size is right
    assert union.size == world.size

    # compute the intersection of the two
    intersection = odds.intersection(evens)
    # verify the this is an empty group
    assert intersection.isEmpty()

    # compute the difference (world - odd)
    difference = whole.difference(odds)
    # verify it is the same size as evens
    assert difference.size == evens.size

    return


# main
if __name__ == "__main__":
    test()


# end of file

File: extern/pyre/tests/libpyre/mpi/group-setops.cc
// -*- coding: utf-8 -*-
//
// michael a.g. aïvázis
// orthologue
// (c) 1998-2019 all rights reserved
//


// for the build system
#include <portinfo>
//  other packages
#include <cassert>
// grab the mpi objects
#include <pyre/mpi.h>

typedef pyre::mpi::group_t group_t;
typedef pyre::mpi::communicator_t communicator_t;

// main program
int main() {
    // initialize
    MPI_Init(0, 0);

    // push down a scope to make sure our local variables get destroyed before MPI_Finalize
    {
        // build a handle to the world communicator
        communicator_t world(MPI_COMM_WORLD, true);
        // get its group
        group_t whole = world.group();
        // compute the size of the world group
        int wsize = whole.size();

        // select a subset of the ranks, say the even ones
        std::vector<int> evens;
        for (int rank = 0; rank < wsize; ++rank) {
            if (rank % 2 == 0) {
                evens.push_back(rank);
            }
        }

        // build the new groups
        group_t even = whole.include(evens);
        group_t odd = whole.exclude(evens);

        // compute the union of these two
        group_t gu = pyre::mpi::groupUnion(even, odd);
        // verify that it is the right size
        assert(gu.size() == world.size());

        // compute the intersection
        group_t gi = pyre::mpi::groupIntersection(even, odd);
        // verify it is empty
        assert(gi.isEmpty());

        // compute the difference world - odd
        group_t gd = pyre::mpi::groupDifference(whole, odd);
        // verify it is the same size as the even group
        assert(gd.size() == even.size());

    }

    // shutdown
    MPI_Finalize();

    // all done
    return 0;
}

// end of file
